kahuna/public/js/components/gr-chips/gr-chips.js

```diff
@@ -0,0 +1,241 @@
+        // Ignore initialisation call
+        if (items !== previousItems) {
+            $grChipsCtrl.onChange();
+        }
+    });
+
+
+    // If clicking on the inner wrapper or the placeholder
+    // (and not in between chips), focus the end of the
+    // component
+    $grChipsCtrl.wrapperClicked = function($event) {
+        if ($event.target.classList.contains('gr-chips__wrapper') ||
+            $event.target.classList.contains('gr-chips__placeholder')) {
+            $grChipsCtrl.focusEndOfLastChip();
+        }
+    };
+
+    $grChipsCtrl.isEmpty = function() {
+        return $grChipsCtrl.items.length === 1 &&
+            $grChipsCtrl.items[0].value === '';
+    };
+
+    $grChipsCtrl.removeChip = function(item) {
+        const index = $grChipsCtrl.items.indexOf(item);
+        if (index !== -1) {
+            removeChipAt(index);
+            normalizeChips();
+        }
+    };
+
+    $grChipsCtrl.removeChipBefore = function(item) {
+        const index = $grChipsCtrl.items.indexOf(item) - 1;
+        if (index >= 0) {
+            removeChipAt(index);
+            normalizeChips();
+        }
+    };
+
+    $grChipsCtrl.removeChipAfter = function(item) {
+        const index = $grChipsCtrl.items.indexOf(item) + 1;
+        if (index <= $grChipsCtrl.items.length) {
+            removeChipAt(index);
+            normalizeChips();
+        }
+    };
+
+    $grChipsCtrl.setFocusedChip = function(item, caretStart = 0, caretEnd = 0) {
+        const index = $grChipsCtrl.items.indexOf(item);
+        if (index !== -1) {
+            $grChipsCtrl.focusedItem = item;
+            $grChipsCtrl.caretStartOffset = caretStart;
+            $grChipsCtrl.caretEndOffset = caretEnd;
+            return true;
+        }
+    };
+
+    $grChipsCtrl.unsetFocusedChip = function(item) {
+        if ($grChipsCtrl.focusedItem === item) {
+            $grChipsCtrl.focusedItem = null;
+            $grChipsCtrl.caretStartOffset = null;
+            $grChipsCtrl.caretEndOffset = null;
+            return true;
+        }
+    };
+
+    $grChipsCtrl.focusEndOfChipBefore = function(item) {
+        const index = $grChipsCtrl.items.indexOf(item) - 1;
+        if (index >= 0) {
+            const previousItem = $grChipsCtrl.items[index];
+            const itemLength = previousItem.value.length;
+            $grChipsCtrl.setFocusedChip(previousItem, itemLength, itemLength);
+            return true;
+        }
+    };
+
+    $grChipsCtrl.focusStartOfChipAfter = function(item) {
+        const index = $grChipsCtrl.items.indexOf(item) + 1;
+        if (index <= $grChipsCtrl.items.length) {
+            $grChipsCtrl.setFocusedChip($grChipsCtrl.items[index], 0, 0);
+            return true;
+        }
+    };
+
+    $grChipsCtrl.focusStartOfFirstChip = function() {
+        const firstItem = $grChipsCtrl.items[0];
+        if (firstItem) {
+            $grChipsCtrl.setFocusedChip(firstItem, 0, 0);
+            return true;
```

Comment:
I've changed it to return the return value of `$grChipsCtrl.setFocusedChip(...)`, as it's more concise, thanks.

The return value is [sometimes used](https://github.com/guardian/grid/pull/1780/files#diff-af864ba9fa711b16b26b24afdfe8af6bR172) to determine whether the focus did something or not, in order to fallback to another behaviour.